sdk/iot/core/tests/cmocka/az_iot_telemetry_tests.c

```diff
@@ -0,0 +1,283 @@
+  }
+#define TEST_VALID_OPTIONS_USER_AGENT \
+  { \
+    .module_id = AZ_SPAN_LITERAL_NULL, .user_agent = AZ_SPAN_LITERAL_FROM_STR(TEST_USER_AGENT) \
+  }
+
+static const az_iot_hub_client_properties g_test_params
+    = { ._internal
+        = { .properties = AZ_SPAN_LITERAL_FROM_STR(TEST_PARAMS), .current_property = 0 } };
+
+static const char g_test_correct_topic_no_options_no_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX;
+static const char g_test_correct_topic_with_options_no_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_USER_AGENT;
+static const char g_test_correct_topic_with_options_with_params[] = TEST_CORRECT_TOPIC_PREFIX
+    "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID "" TEST_MODULE_ID
+    "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS "&" TEST_USER_AGENT;
+static const char g_test_correct_topic_no_options_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_module_id_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_user_agent_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS
+                                "&" TEST_USER_AGENT;
+
+static const az_iot_hub_client g_test_valid_client_no_options
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_EMPTY_OPTIONS } };
+static const az_iot_hub_client g_test_valid_client_with_options_both
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_BOTH } };
+static const az_iot_hub_client g_test_valid_client_with_options_module_id
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_MODULE_ID } };
+static const az_iot_hub_client g_test_valid_client_with_options_user_agent
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_USER_AGENT } };
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_client_fails(void** state)
+{
+  (void)state;
```

Comment:
What's this state and why do we need to be bothered by it?

Reply:
not sure it's a cmocka concept. Can figure it out afterwards if necessary.

Reply:
yeah, this is a cmocka required arg.
It can be used to run tests sharing state. Might be useful when many tests requires the same setup